mods/alt-tab-per-monitor.wh.cpp

```diff
@@ -0,0 +1,189 @@
+                 monInfo.rcWork.bottom - monInfo.rcWork.top,
+                 SWP_NOACTIVATE | SWP_NOZORDER);
+}
+
+void* CWin32ApplicationView_vtable;
+void* CWinRTApplicationView_vtable;
+
+using CWin32ApplicationView_v_GetNativeWindow_t =
+    HRESULT(WINAPI*)(void* pThis, HWND* windowHandle);
+CWin32ApplicationView_v_GetNativeWindow_t
+    CWin32ApplicationView_v_GetNativeWindow;
+
+using CWinRTApplicationView_v_GetNativeWindow_t =
+    HRESULT(WINAPI*)(void* pThis, HWND* windowHandle);
+CWinRTApplicationView_v_GetNativeWindow_t
+    CWinRTApplicationView_v_GetNativeWindow;
+
+using CVirtualDesktop_IsViewVisible_t = HRESULT(WINAPI*)(void* pThis,
+                                                         void* applicationView,
+                                                         BOOL* isVisible);
+CVirtualDesktop_IsViewVisible_t CVirtualDesktop_IsViewVisible_Original;
+HRESULT WINAPI CVirtualDesktop_IsViewVisible_Hook(void* pThis,
+                                                  void* applicationView,
+                                                  BOOL* isVisible) {
+    Wh_Log(L">");
+
+    auto ret = CVirtualDesktop_IsViewVisible_Original(pThis, applicationView,
+                                                      isVisible);
+    if (FAILED(ret)) {
+        return ret;
+    }
+
+    if (!*isVisible) {
+        return ret;
+    }
+
+    void* vtable = *(void**)applicationView;
+
+    HWND windowHandle = nullptr;
+    HRESULT hr = E_FAIL;
+    if (vtable == CWin32ApplicationView_vtable) {
+        hr = CWin32ApplicationView_v_GetNativeWindow(applicationView,
+                                                     &windowHandle);
+    } else if (vtable == CWinRTApplicationView_vtable) {
+        hr = CWinRTApplicationView_v_GetNativeWindow(applicationView,
+                                                     &windowHandle);
+    }
+    if (FAILED(hr) || !windowHandle) {
+        return ret;
+    }
+
+    POINT pt;
+    if (!GetCursorPos(&pt)) {
+        return ret;
+    }
+
+    auto hMon = MonitorFromPoint(pt, MONITOR_DEFAULTTONEAREST);
+
+    auto hMonFromWindow =
+        MonitorFromWindow(windowHandle, MONITOR_DEFAULTTONEAREST);
+
+    if (hMon != hMonFromWindow) {
+        *isVisible = FALSE;
+    }
+
+    return ret;
+}
+
+using XamlAltTabViewHost_DisplayAltTab_t = void(WINAPI*)(void* pThis);
+XamlAltTabViewHost_DisplayAltTab_t XamlAltTabViewHost_DisplayAltTab_Original;
+void XamlAltTabViewHost_DisplayAltTab_Hook(void* pThis) {
+    g_threadIdForAltTabShowWindow = GetCurrentThreadId();
+    XamlAltTabViewHost_DisplayAltTab_Original(pThis);
+    g_threadIdForAltTabShowWindow = 0;
+}
+
+using ShowWindow_t = decltype(&ShowWindow);
+ShowWindow_t ShowWindow_Original;
+BOOL ShowWindow_Hook(HWND hWnd, int nCmdShow) {
+    if (g_threadIdForAltTabShowWindow != GetCurrentThreadId()) {
+        return ShowWindow_Original(hWnd, nCmdShow);
+    }
+
+    if (nCmdShow != SW_HIDE) {
+        HandleAltTabWindow(hWnd);
```

Comment:
To handle the cutting and the flickering, perhaps it's better to hook the window creation. It will also be better for the delay mod.